### PR TITLE
feat: [CDS-48895]: Dynamic provisioning for SSH/WinRm

### PR DESCRIPTION
**Pull request by ivan-83**
Added dynamic provisioning support SSH/WinRm deployments.
- Added the new Provisionable interface which needs to be implemented by all infras which support dynamic provisioning
- Added the new mapper InfrastructureProvisionerMapper where should be placed the logic for dynamic infras provisioning 
- Added static expression evaluator to evaluate host attributes based on provisioners' outputs
- Added support for host properties <+instance.host.properties.<propertyName>>

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41756)

### CODE REVIEW

**Review by acasian-harness — CHANGES_REQUESTED** [review]
> Reviewed 10 of 11 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41756)* status: 11 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

*[a discussion](https://reviewable.io/reviews/harness/harness-core/41756#-NJeulC_CC6GJc_r0jRv:-NJeulCaALLypzSecS6h:bmpqnk8) (no related file):*
@ivan-83 please update the pr description with details what is this change about.
Also is this backwards compatible with old schema? Looking at the code i have some doubts.

___
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureProvisionerMapper.java` line 83 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJevIdx5cTMMBP2HBHM:-NJevIdx5cTMMBP2HBHN:b8p45ww) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureProvisionerMapper.java#L83)):*
> ```Java
>       return pdcInfrastructureOutcome;
>     }
>     throw new InvalidArgumentsException(format("Unknown Infrastructure Kind : [%s]", infrastructure.getKind()));
> ```

Unsupported
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureProvisionerMapper.java` line 95 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJewnjODinGpxdDF92B:-NJewnjODinGpxdDF92C:bz602yn) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureProvisionerMapper.java#L95)):*
> ```Java
>     }
> 
>     if (!hasValueOrExpression(infrastructure.getHostAttributes(), false)) {
> ```

old pdc infra that don't have it will start to break
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureProvisionerMapper.java` line 107 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJexvwJ6beqbcr5oMk9:-NJexvwJ6beqbcr5oMkA:bv4u5ma) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureProvisionerMapper.java#L107)):*
> ```Java
>     try {
>       TypeReference<List<Map<String, Object>>> typeRef = new TypeReference<>() {};
>       return new ObjectMapper().readValue(IOUtils.toInputStream(hostInstancesJson), typeRef);
> ```

that `hostInstancesJson` could be a huge file, have we considered the performance metrics? Also there could parsing exceptions, why are we considering only IO ones?
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/AbstractInfrastructureTaskExecutableStep.java` line 188 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJeyT3tDzF58y1eLRkG:-NJeyT3tDzF58y1eLRkH:b-9f16qn) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/AbstractInfrastructureTaskExecutableStep.java#L188)):*
> ```Java
> 
>     final InfrastructureOutcome infrastructureOutcome =
>         (infrastructure instanceof Provisionable && ((Provisionable) infrastructure).isDynamicallyProvisioned())
> ```

old infra won't have this flag
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureStep.java` line 168 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJeyZJc2-wSiZR1lqhg:-NJeyZJc2-wSiZR1lqhh:bb7od0r) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureStep.java#L168)):*
> ```Java
> 
>     InfrastructureOutcome infrastructureOutcome =
>         (infrastructure instanceof Provisionable && ((Provisionable) infrastructure).isDynamicallyProvisioned())
> ```

same
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2.java` line 324 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJeyaq28l4TzztPcOEG:-NJeyaq28l4TzztPcOEH:bb7od0r) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2.java#L324)):*
> ```Java
> 
>     final InfrastructureOutcome infrastructureOutcome =
>         (spec instanceof Provisionable && ((Provisionable) spec).isDynamicallyProvisioned())
> ```

same
___
*[`125-cd-nextgen/src/test/java/io/harness/cdng/infra/InfrastructureProvisionerMapperTest.java` line 46 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJezdtq2qCeRZDvsmcg:-NJezdtq2qCeRZDvsmch:b-dpyi4e) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/test/java/io/harness/cdng/infra/InfrastructureProvisionerMapperTest.java#L46)):*
> ```Java
> @OwnedBy(HarnessTeam.CDC)
> @RunWith(MockitoJUnitRunner.class)
> public class InfrastructureProvisionerMapperTest extends CategoryTest {
> ```

Can we have more coverage on the use cases
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/infra/beans/PdcInfraMapping.java` line 39 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJeywiz3LPU2xHVFgeB:-NJeywiz3LPU2xHVFgeC:b-i8hhv9) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/127-cd-nextgen-entities/src/main/java/io/harness/cdng/infra/beans/PdcInfraMapping.java#L39)):*
> ```Java
>   private Boolean dynamicallyProvisioned;
>   private String hostObjectArray;
>   private Map<String, String> hostAttributes;
> ```

shall we use a dedicated type for provisioning? Is this going to be reused by other infra?

**Review by ivan-83 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41756)* status: 11 file(s) to review. 1 change requests. 8. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureProvisionerMapper.java` line 83 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJevIdx5cTMMBP2HBHM:-NJt8N4JCOub5jscFmUx:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureProvisionerMapper.java#L83)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

Unsupported
</blockquote></details>

Done.
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureProvisionerMapper.java` line 95 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJewnjODinGpxdDF92B:-NJt5fqB8JfkMGlg4Nuh:buscl5d) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureProvisionerMapper.java#L95)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

old pdc infra that don't have it will start to break
</blockquote></details>

This part of the code will be only executed for provisioned infra, this is the provisioner mapper and the logic within this class will be consumed if the dynamic infrastructure mapping radio button is selected on UI. 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureProvisionerMapper.java` line 107 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJexvwJ6beqbcr5oMk9:-NJt9CJEDUYpwlu0_oor:b-4dtp5d) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureProvisionerMapper.java#L107)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

that `hostInstancesJson` could be a huge file, have we considered the performance metrics? Also there could parsing exceptions, why are we considering only IO ones?
</blockquote></details>

IO exception is only checked one and needs additional handling for it. The other exceptions are unchecked and will be handled by our framework.



_Code quote:_
```Java
parseHostInstancesJSON
```
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/AbstractInfrastructureTaskExecutableStep.java` line 188 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJeyT3tDzF58y1eLRkG:-NJt6XQzCL7MHmCOXNFL:bmdv2c8) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/AbstractInfrastructureTaskExecutableStep.java#L188)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

old infra won't have this flag
</blockquote></details>

Right, in order to resolve this isDynamicallyProvisioned() we are calling this method,
ParameterFieldHelper.getBooleanParameterFieldValue(dynamicallyProvisioned)  where we have this part of the code 

    if (value == null) {
      return Boolean.FALSE;
    }

and the above part of the code solves that scenario. 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureStep.java` line 168 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJeyZJc2-wSiZR1lqhg:-NJt7a941r-7vfoGJ5n6:b-ia41wb) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureStep.java#L168)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

same
</blockquote></details>

same comment as above
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2.java` line 324 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJeyaq28l4TzztPcOEG:-NJt7doS0oZ4_Ecte0wZ:b1jeszp) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2.java#L324)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

same
</blockquote></details>

same comment as above
___
*[`125-cd-nextgen/src/test/java/io/harness/cdng/infra/InfrastructureProvisionerMapperTest.java` line 46 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJezdtq2qCeRZDvsmcg:-NJtCCT28B3GRRBCr8BT:b2nign6) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/125-cd-nextgen/src/test/java/io/harness/cdng/infra/InfrastructureProvisionerMapperTest.java#L46)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

Can we have more coverage on the use cases
</blockquote></details>

yes
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/infra/beans/PdcInfraMapping.java` line 39 at r2](https://reviewable.io/reviews/harness/harness-core/41756#-NJeywiz3LPU2xHVFgeB:-NJt7s8bC-BXfkTa6DFj:b-b42tu3) ([raw file](https://github.com/harness/harness-core/blob/3b055b04abd1362025700c0e06877fc8cba749e4/127-cd-nextgen-entities/src/main/java/io/harness/cdng/infra/beans/PdcInfraMapping.java#L39)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

shall we use a dedicated type for provisioning? Is this going to be reused by other infra?
</blockquote></details>

hostObjectArray and hostAttributes will be only used by PDC infra and we don't have similar logic for other infras.  Provisionable interface should be used by other infras where we have only one common method  isDynamicallyProvisioned() .

**Comment by ivan-83** [comment]
trigger messagecheck

**Comment by ivan-83** [comment]
trigger checkstylecodeformat

**Review by acasian-harness — DISMISSED** [review]
> Reviewed 9 of 10 files at r3, 5 of 5 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41756)* status: 19 file(s) to review. 1 of 2 approvals obtained.

File: 125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2.java [hunk 1]
```diff
@@ -317,8 +318,11 @@ private void executeSync(
 
     infrastructureValidator.validate(spec);
 
-    final InfrastructureOutcome infrastructureOutcome = infrastructureMapper.toOutcome(spec, environmentOutcome,
-        serviceOutcome, ngAccess.getAccountIdentifier(), ngAccess.getOrgIdentifier(), ngAccess.getProjectIdentifier());
+    final InfrastructureOutcome infrastructureOutcome =
+        (spec instanceof Provisionable && ((Provisionable) spec).isDynamicallyProvisioned())
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2.java:322]: Can we use factory pattern here 

outcomeGenerator = InfraOutcomeFactory.getOutcomeGenerator(..);
outcome = outcomeGenerate.generate(..) ?

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/expressionEvaluator/CDEngineExpressionEvaluator.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * The purpose of CDEngineExpressionEvaluator is to dynamically evaluate expressions based on the context that is
+ * supplied to methods. This evaluator cannot evaluate PMS expressions because it doesn't make calls to PMS service. If
+ * you need to evaluate PMS expressions use {@link io.harness.cdng.expressions.CDExpressionResolver}
+ */
+@OwnedBy(HarnessTeam.CDP)
+@Slf4j
+public class CDEngineExpressionEvaluator extends EngineExpressionEvaluator {
```
Comment (yogesh-chauhan) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/expressionEvaluator/CDEngineExpressionEvaluator.java:28]: can we maybe call it CDLocalExpressionEvaluator..

Comment (yogesh-chauhan) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/expressionEvaluator/CDEngineExpressionEvaluator.java:28]: do we need to extend EngineExpressionEvaluator

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

**Review by ivan-83 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41756)* status: 23 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2.java` line 322 at r8](https://reviewable.io/reviews/harness/harness-core/41756#-NJeazDUn_T_NZapnHKj-r8-322:-NLRjIgm2wvezRTeC0Jw:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/eaaed00eef0de7150c636b96b1ec5abab8b28afb/125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2.java#L322)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

Can we use factory pattern here 

outcomeGenerator = InfraOutcomeFactory.getOutcomeGenerator(..);
outcome = outcomeGenerate.generate(..) ?

</blockquote></details>

Done
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/expressionEvaluator/CDEngineExpressionEvaluator.java` line 28 at r8](https://reviewable.io/reviews/harness/harness-core/41756#-NJeazDYkEfoxWZasmq8-r8-28:-NLRjTHKCfhBI01uVjYg:b2svfja) ([raw file](https://github.com/harness/harness-core/blob/eaaed00eef0de7150c636b96b1ec5abab8b28afb/127-cd-nextgen-entities/src/main/java/io/harness/cdng/expressionEvaluator/CDEngineExpressionEvaluator.java#L28)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we maybe call it CDLocalExpressionEvaluator..
</blockquote></details>

CDExpressionEvaluator  has already been added in meantime and reused it.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 11 files at r1, 3 of 10 files at r3, 1 of 5 files at r4, 1 of 1 files at r5, 2 of 7 files at r6, 1 of 2 files at r7, 4 of 4 files at r8, 10 of 12 files at r9, 3 of 3 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41756)* status: 18 file(s) to review. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 4 of 4 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41756)* status: 18 file(s) to review. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41756)* status: 18 file(s) to review. 1 of 2 approvals obtained.

**Review by acasian-harness — APPROVED** [review]
> Reviewed 1 of 7 files at r6, 3 of 4 files at r8, 9 of 12 files at r9, 1 of 3 files at r10, 2 of 4 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41756)* status: 2 file(s) to review.

**Review by tchaurasiya — APPROVED** [review]
> Reviewed 1 of 11 files at r1, 3 of 10 files at r3, 1 of 5 files at r4, 1 of 1 files at r5, 1 of 7 files at r6, 1 of 2 files at r7, 3 of 4 files at r8, 9 of 12 files at r9, 1 of 3 files at r10, 4 of 4 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41756)* status: :shipit: complete! ✓

**Pull request merged**